Feat: Adding Yamux as default multiplexer, keeping Mplex as fallback

paschal533 opened the pull request:
## Description

This PR replaces the deprecated mplex multiplexer with yamux in py-libp2p, aligning it with modern libp2p implementations like rust-libp2p and go-libp2p. yamux offers improved performance with flow control and backpressure, enhancing interoperability.

#### Changes

- Added yamux.py in libp2p/stream_muxer/yamux/ with a full implementation supporting stream creation, closure, and reset per the Yamux spec.

- Updated libp2p/__init__.py:

  - Replaced mplex with yamux as the default in muxer_transports_by_protocol.

  - Switched default key generation to Ed25519 from RSA for better security.

  - Set Noise as the default security transport over insecure/secio.
 
- Updated libp2p/tools/factories.py

- Updated tests/conftest.py

- Updated tests/core/stream_muxer/conftest.py

- Adjusted examples/ping/ping.py to use yamux (separate commit if needed).

- Fixed type annotations and imports for compatibility.

#### Related Issue

- Addresses #534 

#### Testing

- Created `test_yamux.py` in `tests/core/stream_muxer` folder

- Manual test with examples/ping/ping.py confirms yamux usage.

- Interop test with rust-libp2p pending Noise handshake resolution (tracked separately).

#### Notes

- mplex is removed as the default but can be re-added as a fallback if maintainers prefer gradual deprecation.

#### Request for Feedback

- Should mplex remain as an optional fallback?

pacrob commented:
> * Should mplex remain as an optional fallback?

It looks like mplex has been retained, sometimes in a deprecated status, in other implementations. Let's keep it for now, with some clear messaging that yamux is preferred.

paschal533 commented:
> > * Should mplex remain as an optional fallback?
> 
> It looks like mplex has been retained, sometimes in a deprecated status, in other implementations. Let's keep it for now, with some clear messaging that yamux is preferred.

Okay

paschal533 commented:
Hi @pacrob. Thanks for the feedback. I’ve ensured that `mplex` remains untouched in the codebase and added clear messaging in   `yamux.py` comments to indicate that Yamux is preferred, while Mplex is retained for compatibility. I’ve also fixed the `mypy` errors in `YamuxStream` related to `get_remote_address`. Let me know if there’s anything else you’d like adjusted

pacrob commented on libp2p/stream_muxer/yamux/yamux.py:
```diff
@@ -0,0 +1,270 @@
+    async def write(self, data: bytes) -> None:
+        if self.closed:
+            raise MuxedStreamError("Stream is closed")
+        header = struct.pack("!BBHII", 0, TYPE_DATA, 0, self.stream_id, len(data))
```
"!BBHII" looks like a random value here, and it's used multiple times in this file. Could you pull it up with the other constants at the top?

paschal533 replied:
Thanks for the suggestion. I’ve moved `"!BBHII"` to a constant `YAMUX_HEADER_FORMAT` at the top of `yamux.py` with a comment explaining its structure. It’s now used consistently in `write`, `close`, and `reset`.

pacrob replied:
looks good.

pacrob commented on newsfragments/534.bugfix.rst:
I think this is more of a `feature` than a `bugfix`.

paschal533 replied:
You’re right, adding Yamux is more of a feature than a bugfix. I’ve renamed the news fragment to `534.feature.rst` and updated the description to: "Added support for the Yamux stream multiplexer (/yamux/1.0.0) as the preferred option, retaining Mplex (/mplex/6.7.0) for backward compatibility." Does that look good?

pacrob replied:
great!

pacrob commented on libp2p/stream_muxer/yamux/yamux.py:
```diff
@@ -0,0 +1,270 @@
+            await self.conn.secured_conn.write(header)
+            self.closed = True
+
+    def set_deadline(self, ttl: int) -> bool:
```
What does this do, that it should always return `False`?

paschal533 replied:
The `set_deadline` method returns `False` because Yamux doesn’t natively support setting deadlines (unlike Mplex, which tracks them). I’ve added a docstring to clarify this: "Yamux does not support deadlines natively, so this method always returns False." If we need deadline support, we could extend it later, but for now, this aligns with the spec. Thoughts?

pacrob replied:
The comment is good, but for an end user, I think raising a `NotImplementedError` is probably more useful.

paschal533 replied:
Alright, I’ve implemented the change to raise `NotImplementedError` in `set_deadline`. I hit an `AttributeError: 'YamuxStream' object has no attribute 'muxed_conn'` when adding a test for `set_deadline`, due to `NetStream` expecting `muxed_conn`. I fixed it by adding `self.muxed_conn = conn` to `YamuxStream.__init__`, consistent with `MplexStream`. I verify it raises `NotImplementedError`. All `yamux` and `mplex` tests now pass.

pacrob commented on tests/core/stream_muxer/test_yamux.py:
```diff
@@ -0,0 +1,195 @@
+    print("test_yamux_connection_close complete")
+
+
+if __name__ == "__main__":
```
I don't think we need this in a test file.

paschal533 replied:
Agreed. I’ve removed it from `test_yamux.py`. Thanks for pointing that out.

pacrob commented on tests/core/stream_muxer/test_yamux.py:
```diff
@@ -0,0 +1,195 @@
+        self.receive_stream = receive_stream
+
+    async def write(self, data):
+        print(f"Writing {len(data)} bytes")
```
`print` statements are fine during development, but we'd prefer to use logging in production. That way the output doesn't clutter up CI logs, but can be turned on when running pytest locally.

paschal533 replied:
Totally agree on using `logging` over `print` for production readiness. I’ve replaced the `print` statements in `test_yamux.py` with `logging.debug`.

pacrob commented on libp2p/stream_muxer/yamux/yamux.py:
```diff
@@ -0,0 +1,270 @@
+        self.is_initiator_value = (
+            is_initiator if is_initiator is not None else secured_conn.is_initiator
+        )
+        self.next_stream_id = 1 if self.is_initiator else 2
```
I was curious about this. A comment referring to how this is from yamux spec def for stream_id would be helpful.

pacrob replied:
great!

seetadev commented:
@Khwahish29 , @acul71 This is an important feature enhancement. @pacrob has shared key pointers and feedback. Please study it and discuss with @paschal533, @pacrob and me. Would like you to contribute to this core issue.   

Lets catch up and arrive at a good conclusion to the feature enhancement by Thursday. We will review it during maintainers' call.

pacrob commented:
Just the one request for raising an error, otherwise everything I noted looks good now. It looks like you're still working on this, so let me know when it's ready for another look.

paschal533 commented:
> Just the one request for raising an error, otherwise everything I noted looks good now. It looks like you're still working on this, so let me know when it's ready for another look.

Alright Paul, I will let you know once it's ready for another look. Thanks for the reviews.

acul71 commented:
@paschal533 I can see this tests are failing:
I can work on those, but let me know if you want to do it, so we don't work on the same code twice 
```bash
================================================================== short test summary info ===================================================================
FAILED tests/core/host/test_connected_peers.py::test_connected_peers[connect_and_disconnect] - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/core/host/test_live_peers.py::test_live_peers_disconnect - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/core/host/test_live_peers.py::test_live_peers_unexpected_drop - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/core/identity/identify/test_identify.py::test_identify_protocol - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/core/network/test_net_stream.py::test_net_stream_read_until_eof - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/core/network/test_net_stream.py::test_net_stream_read_after_remote_closed - BaseExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/core/network/test_net_stream.py::test_net_stream_read_after_local_reset - BaseExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/core/network/test_net_stream.py::test_net_stream_read_after_remote_reset - BaseExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/core/network/test_net_stream.py::test_net_stream_write_after_local_closed - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/core/network/test_net_stream.py::test_net_stream_write_after_local_reset - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 10 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=============================================================== 10 failed, 28 passed in 17.99s ===============================================================
```
and others too.

# Yamux Implementation Analysis and Test Failures

## Overview
This document analyzes the current Yamux implementation and the failing tests in the codebase. The analysis focuses on stream handling, connection management, and the specific issues causing test failures.

## Test Failures Summary
The following tests are currently failing:
- `test_connected_peers[connect_and_disconnect]`
- `test_live_peers_disconnect`
- `test_live_peers_unexpected_drop`
- `test_identify_protocol`
- Multiple `test_net_stream` tests:
  - `test_net_stream_read_until_eof`
  - `test_net_stream_read_after_remote_closed`
  - `test_net_stream_read_after_local_reset`
  - `test_net_stream_read_after_remote_reset`
  - `test_net_stream_write_after_local_closed`
  - `test_net_stream_write_after_local_reset`

## Identified Issues

### 1. Stream Closure Handling
- The `close()` method in `Yamux` sets `stream.closed = True` but doesn't properly notify waiting readers/writers
- The arbitrary `trio.sleep(0.1)` in the close method may cause race conditions
- Stream closure events are not properly propagated to all interested parties

### 2. Stream Reset Handling
- RST flag handling in the Yamux implementation doesn't properly propagate the reset state
- Stream reset events may not be properly cleaning up resources
- Reset conditions may not be properly signaled to waiting operations

### 3. EOF Handling
- EOF conditions aren't being properly signaled
- FIN flag handling may not be properly implemented
- Waiting readers may not be properly notified of EOF conditions

### 4. Connection Shutdown
- GO_AWAY handling in `handle_incoming` may not be properly cleaning up all resources
- Connection cleanup issues in peer disconnection scenarios
- Resource leaks during unexpected connection drops

## Implementation Issues

### 1. Close Method Issues
```python
async def close(self) -> None:
    # ...
    await trio.sleep(0.1)  # Arbitrary sleep causing potential issues
```

### 2. Stream Event Handling
```python
if flags & FLAG_FIN:
    logging.debug(f"Closing stream {stream_id} due to FIN")
    self.streams[stream_id].closed = True
    # Missing: Should set the stream event to wake up any waiting readers
```

### 3. Reset Handling
```python
elif typ == TYPE_DATA and flags & FLAG_RST:
    async with self.streams_lock:
        if stream_id in self.streams:
            logging.debug(f"Resetting stream {stream_id}")
            self.streams[stream_id].closed = True
            self.stream_events[stream_id].set()
```

## Proposed Solutions

### 1. Stream Closure Improvements
- Remove the arbitrary sleep in the close method
- Implement proper event notification for stream closure
- Ensure all waiting operations are properly notified of closure

### 2. Reset Handling Enhancements
- Improve RST flag handling to properly propagate reset state
- Implement proper cleanup of resources during reset
- Add proper error signaling for reset conditions

### 3. EOF Handling Implementation
- Properly implement FIN flag handling
- Add proper EOF signaling to waiting readers
- Ensure EOF conditions are properly propagated

### 4. Connection Shutdown Improvements
- Enhance GO_AWAY handling to ensure proper resource cleanup
- Implement proper connection cleanup procedures
- Add proper error handling for unexpected disconnections

## Next Steps
1. Remove the arbitrary sleep in the close method
2. Implement proper stream event propagation
3. Enhance reset and EOF handling
4. Improve connection shutdown procedures
5. Add comprehensive error handling
6. Update tests to verify the fixes

## Conclusion
The current implementation has several issues that need to be addressed to ensure proper stream handling and connection management. The proposed solutions focus on improving event propagation, resource cleanup, and error handling to resolve the failing tests and improve overall reliability.

acul71 commented:
@paschal533 

# Yamux Specification Compliance Analysis

## Overview
This document analyzes the current Yamux implementation against the official libp2p Yamux specification. The analysis identifies implemented features, missing components, and areas requiring improvement.

## Specification Compliance Analysis

### 1. Protocol String
- **Specification**: `/yamux/1.0.0`
- **Implementation Status**: ✅ Implemented
- **Details**: Correctly implemented as `PROTOCOL_ID = "/yamux/1.0.0"`

### 2. Frame Structure
- **Specification**: 12-byte header with:
  - Version (8 bits)
  - Type (8 bits)
  - Flags (16 bits)
  - StreamID (32 bits)
  - Length (32 bits)
- **Implementation Status**: ✅ Implemented
- **Details**: Correctly implemented with `YAMUX_HEADER_FORMAT = "!BBHII"`

### 3. Message Types
- **Specification**: 
  - 0x0 Data
  - 0x1 Window Update
  - 0x2 Ping
  - 0x3 Go Away
- **Implementation Status**: ⚠️ Partially Implemented
- **Details**:
  - ✅ Data (TYPE_DATA = 0x0)
  - ✅ Window Update (TYPE_WINDOW_UPDATE = 0x1)
  - ✅ Ping (TYPE_PING = 0x2)
  - ✅ Go Away (TYPE_GO_AWAY = 0x3)
  - ❌ Missing: Window Update functionality

### 4. Flags
- **Specification**:
  - 0x1 SYN
  - 0x2 ACK
  - 0x4 FIN
  - 0x8 RST
- **Implementation Status**: ✅ Implemented
- **Details**: Correctly implemented with:
  ```python
  FLAG_SYN = 0x1
  FLAG_ACK = 0x2
  FLAG_FIN = 0x4
  FLAG_RST = 0x8
  ```

### 5. StreamID Allocation
- **Specification**: Client uses odd IDs, server uses even IDs
- **Implementation Status**: ✅ Implemented
- **Details**: Correctly implemented:
  ```python
  self.next_stream_id = 1 if self.is_initiator else 2
  ```

### 6. Stream Opening
- **Specification**: Initial frame with SYN flag, receiver replies with ACK or RST
- **Implementation Status**: ⚠️ Partially Implemented
- **Details**:
  - ✅ SYN flag handling
  - ❌ Missing: Proper ACK handling for new streams
  - ❌ Missing: RST handling for stream rejection

### 7. Stream Closing
- **Specification**: FIN flag for half-close, RST for immediate close
- **Implementation Status**: ⚠️ Partially Implemented
- **Details**:
  - ✅ FIN flag handling
  - ✅ RST flag handling
  - ❌ Missing: Proper half-close state management

### 8. Flow Control
- **Specification**: Initial 256KiB window size, window updates for flow control
- **Implementation Status**: ❌ Missing
- **Details**:
  - No window size tracking
  - No window update mechanism
  - No flow control implementation

### 9. Session Termination
- **Specification**: GO_AWAY message with error codes:
  - 0x0 Normal termination
  - 0x1 Protocol error
  - 0x2 Internal error
- **Implementation Status**: ⚠️ Partially Implemented
- **Details**:
  - ✅ GO_AWAY message sending
  - ❌ Missing: Error code handling
  - ❌ Missing: Proper session cleanup

### 10. ACK Backlog & Backpressure
- **Specification**: Maximum 256 stream ACK backlog
- **Implementation Status**: ❌ Missing
- **Details**:
  - No ACK backlog tracking
  - No backpressure mechanism

## Major Missing Features

### 1. Flow Control
- Window size management
- Window update mechanism
- Flow control enforcement

### 2. Stream Management
- Proper ACK handling for new streams
- Stream rejection with RST
- Half-close state management

### 3. Error Handling
- GO_AWAY error codes
- Protocol error handling
- Internal error handling

### 4. Backpressure
- ACK backlog management
- Stream opening backpressure
- Resource management

## Recommendations

### 1. Flow Control Implementation
- Add window size tracking
- Implement window update mechanism
- Add flow control enforcement

### 2. Stream Management Improvements
- Add proper ACK handling
- Implement stream rejection
- Add half-close state tracking

### 3. Error Handling Enhancements
- Add error code handling
- Implement proper error propagation
- Add error recovery mechanisms

### 4. Backpressure Implementation
- Implement ACK backlog tracking
- Add stream opening backpressure
- Improve resource management

## Conclusion
The current implementation is missing several critical features from the specification, particularly around flow control and proper stream management. These missing features are likely contributing to the test failures observed in the codebase. Implementing the missing features according to the specification should resolve the current issues and improve the overall reliability of the implementation.

paschal533 commented:
> @paschal533 I can see this tests are failing: I can work on those, but let me know if you want to do it, so we don't work on the same code twice
> 
> ```shell
> ================================================================== short test summary info ===================================================================
> FAILED tests/core/host/test_connected_peers.py::test_connected_peers[connect_and_disconnect] - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
> FAILED tests/core/host/test_live_peers.py::test_live_peers_disconnect - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
> FAILED tests/core/host/test_live_peers.py::test_live_peers_unexpected_drop - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
> FAILED tests/core/identity/identify/test_identify.py::test_identify_protocol - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
> FAILED tests/core/network/test_net_stream.py::test_net_stream_read_until_eof - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
> FAILED tests/core/network/test_net_stream.py::test_net_stream_read_after_remote_closed - BaseExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
> FAILED tests/core/network/test_net_stream.py::test_net_stream_read_after_local_reset - BaseExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
> FAILED tests/core/network/test_net_stream.py::test_net_stream_read_after_remote_reset - BaseExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
> FAILED tests/core/network/test_net_stream.py::test_net_stream_write_after_local_closed - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
> FAILED tests/core/network/test_net_stream.py::test_net_stream_write_after_local_reset - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
> !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 10 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
> =============================================================== 10 failed, 28 passed in 17.99s ===============================================================
> ```
> 
> and others too.
> 
> # Yamux Implementation Analysis and Test Failures
> ## Overview
> This document analyzes the current Yamux implementation and the failing tests in the codebase. The analysis focuses on stream handling, connection management, and the specific issues causing test failures.
> 
> ## Test Failures Summary
> The following tests are currently failing:
> 
> * `test_connected_peers[connect_and_disconnect]`
> * `test_live_peers_disconnect`
> * `test_live_peers_unexpected_drop`
> * `test_identify_protocol`
> * Multiple `test_net_stream` tests:
>   
>   * `test_net_stream_read_until_eof`
>   * `test_net_stream_read_after_remote_closed`
>   * `test_net_stream_read_after_local_reset`
>   * `test_net_stream_read_after_remote_reset`
>   * `test_net_stream_write_after_local_closed`
>   * `test_net_stream_write_after_local_reset`
> 
> ## Identified Issues
> ### 1. Stream Closure Handling
> * The `close()` method in `Yamux` sets `stream.closed = True` but doesn't properly notify waiting readers/writers
> * The arbitrary `trio.sleep(0.1)` in the close method may cause race conditions
> * Stream closure events are not properly propagated to all interested parties
> 
> ### 2. Stream Reset Handling
> * RST flag handling in the Yamux implementation doesn't properly propagate the reset state
> * Stream reset events may not be properly cleaning up resources
> * Reset conditions may not be properly signaled to waiting operations
> 
> ### 3. EOF Handling
> * EOF conditions aren't being properly signaled
> * FIN flag handling may not be properly implemented
> * Waiting readers may not be properly notified of EOF conditions
> 
> ### 4. Connection Shutdown
> * GO_AWAY handling in `handle_incoming` may not be properly cleaning up all resources
> * Connection cleanup issues in peer disconnection scenarios
> * Resource leaks during unexpected connection drops
> 
> ## Implementation Issues
> ### 1. Close Method Issues
> ```python
> async def close(self) -> None:
>     # ...
>     await trio.sleep(0.1)  # Arbitrary sleep causing potential issues
> ```
> 
> ### 2. Stream Event Handling
> ```python
> if flags & FLAG_FIN:
>     logging.debug(f"Closing stream {stream_id} due to FIN")
>     self.streams[stream_id].closed = True
>     # Missing: Should set the stream event to wake up any waiting readers
> ```
> 
> ### 3. Reset Handling
> ```python
> elif typ == TYPE_DATA and flags & FLAG_RST:
>     async with self.streams_lock:
>         if stream_id in self.streams:
>             logging.debug(f"Resetting stream {stream_id}")
>             self.streams[stream_id].closed = True
>             self.stream_events[stream_id].set()
> ```
> 
> ## Proposed Solutions
> ### 1. Stream Closure Improvements
> * Remove the arbitrary sleep in the close method
> * Implement proper event notification for stream closure
> * Ensure all waiting operations are properly notified of closure
> 
> ### 2. Reset Handling Enhancements
> * Improve RST flag handling to properly propagate reset state
> * Implement proper cleanup of resources during reset
> * Add proper error signaling for reset conditions
> 
> ### 3. EOF Handling Implementation
> * Properly implement FIN flag handling
> * Add proper EOF signaling to waiting readers
> * Ensure EOF conditions are properly propagated
> 
> ### 4. Connection Shutdown Improvements
> * Enhance GO_AWAY handling to ensure proper resource cleanup
> * Implement proper connection cleanup procedures
> * Add proper error handling for unexpected disconnections
> 
> ## Next Steps
> 1. Remove the arbitrary sleep in the close method
> 2. Implement proper stream event propagation
> 3. Enhance reset and EOF handling
> 4. Improve connection shutdown procedures
> 5. Add comprehensive error handling
> 6. Update tests to verify the fixes
> 
> ## Conclusion
> The current implementation has several issues that need to be addressed to ensure proper stream handling and connection management. The proposed solutions focus on improving event propagation, resource cleanup, and error handling to resolve the failing tests and improve overall reliability.

Yeah, sure, Luca. You can work on the test. 😊

paschal533 commented:
> @paschal533
> 
> # Yamux Specification Compliance Analysis
> ## Overview
> This document analyzes the current Yamux implementation against the official libp2p Yamux specification. The analysis identifies implemented features, missing components, and areas requiring improvement.
> 
> ## Specification Compliance Analysis
> ### 1. Protocol String
> * **Specification**: `/yamux/1.0.0`
> * **Implementation Status**: ✅ Implemented
> * **Details**: Correctly implemented as `PROTOCOL_ID = "/yamux/1.0.0"`
> 
> ### 2. Frame Structure
> * **Specification**: 12-byte header with:
>   
>   * Version (8 bits)
>   * Type (8 bits)
>   * Flags (16 bits)
>   * StreamID (32 bits)
>   * Length (32 bits)
> * **Implementation Status**: ✅ Implemented
> * **Details**: Correctly implemented with `YAMUX_HEADER_FORMAT = "!BBHII"`
> 
> ### 3. Message Types
> * **Specification**:
>   
>   * 0x0 Data
>   * 0x1 Window Update
>   * 0x2 Ping
>   * 0x3 Go Away
> * **Implementation Status**: ⚠️ Partially Implemented
> * **Details**:
>   
>   * ✅ Data (TYPE_DATA = 0x0)
>   * ✅ Window Update (TYPE_WINDOW_UPDATE = 0x1)
>   * ✅ Ping (TYPE_PING = 0x2)
>   * ✅ Go Away (TYPE_GO_AWAY = 0x3)
>   * ❌ Missing: Window Update functionality
> 
> ### 4. Flags
> * **Specification**:
>   
>   * 0x1 SYN
>   * 0x2 ACK
>   * 0x4 FIN
>   * 0x8 RST
> * **Implementation Status**: ✅ Implemented
> * **Details**: Correctly implemented with:
>   ```python
>   FLAG_SYN = 0x1
>   FLAG_ACK = 0x2
>   FLAG_FIN = 0x4
>   FLAG_RST = 0x8
>   ```
> 
> ### 5. StreamID Allocation
> * **Specification**: Client uses odd IDs, server uses even IDs
> * **Implementation Status**: ✅ Implemented
> * **Details**: Correctly implemented:
>   ```python
>   self.next_stream_id = 1 if self.is_initiator else 2
>   ```
> 
> ### 6. Stream Opening
> * **Specification**: Initial frame with SYN flag, receiver replies with ACK or RST
> * **Implementation Status**: ⚠️ Partially Implemented
> * **Details**:
>   
>   * ✅ SYN flag handling
>   * ❌ Missing: Proper ACK handling for new streams
>   * ❌ Missing: RST handling for stream rejection
> 
> ### 7. Stream Closing
> * **Specification**: FIN flag for half-close, RST for immediate close
> * **Implementation Status**: ⚠️ Partially Implemented
> * **Details**:
>   
>   * ✅ FIN flag handling
>   * ✅ RST flag handling
>   * ❌ Missing: Proper half-close state management
> 
> ### 8. Flow Control
> * **Specification**: Initial 256KiB window size, window updates for flow control
> * **Implementation Status**: ❌ Missing
> * **Details**:
>   
>   * No window size tracking
>   * No window update mechanism
>   * No flow control implementation
> 
> ### 9. Session Termination
> * **Specification**: GO_AWAY message with error codes:
>   
>   * 0x0 Normal termination
>   * 0x1 Protocol error
>   * 0x2 Internal error
> * **Implementation Status**: ⚠️ Partially Implemented
> * **Details**:
>   
>   * ✅ GO_AWAY message sending
>   * ❌ Missing: Error code handling
>   * ❌ Missing: Proper session cleanup
> 
> ### 10. ACK Backlog & Backpressure
> * **Specification**: Maximum 256 stream ACK backlog
> * **Implementation Status**: ❌ Missing
> * **Details**:
>   
>   * No ACK backlog tracking
>   * No backpressure mechanism
> 
> ## Major Missing Features
> ### 1. Flow Control
> * Window size management
> * Window update mechanism
> * Flow control enforcement
> 
> ### 2. Stream Management
> * Proper ACK handling for new streams
> * Stream rejection with RST
> * Half-close state management
> 
> ### 3. Error Handling
> * GO_AWAY error codes
> * Protocol error handling
> * Internal error handling
> 
> ### 4. Backpressure
> * ACK backlog management
> * Stream opening backpressure
> * Resource management
> 
> ## Recommendations
> ### 1. Flow Control Implementation
> * Add window size tracking
> * Implement window update mechanism
> * Add flow control enforcement
> 
> ### 2. Stream Management Improvements
> * Add proper ACK handling
> * Implement stream rejection
> * Add half-close state tracking
> 
> ### 3. Error Handling Enhancements
> * Add error code handling
> * Implement proper error propagation
> * Add error recovery mechanisms
> 
> ### 4. Backpressure Implementation
> * Implement ACK backlog tracking
> * Add stream opening backpressure
> * Improve resource management
> 
> ## Conclusion
> The current implementation is missing several critical features from the specification, particularly around flow control and proper stream management. These missing features are likely contributing to the test failures observed in the codebase. Implementing the missing features according to the specification should resolve the current issues and improve the overall reliability of the implementation.

Thanks Luca. I will work on this.

seetadev commented:
@acul71 — thank you for sharing the test results and your analysis of the PR. Really appreciate the thoroughness.

I'd like you and @paschal533 to coordinate on wrapping up this PR in time for the review scheduled on Thursday. Please make sure all pending items are addressed and the final state is review-ready by then.

To keep things organized, let’s continue the discussion over at: https://github.com/libp2p/py-libp2p/discussions/568.
This will help consolidate all context and updates in one place, separate from the PR thread.

@dhuseby and @pacrob : Using the discussion page will help us keep the PR page clean and focused on the code diffs, reviews & comments, and final approvals.

acul71 commented:
@paschal533 @seetadev 
I've fixed 2 failing tests for `Yamux`
See [here](https://github.com/paschal533/py-libp2p/pull/1)

In total 4 tests are failing now only this two are still failing:

```bash
FAILED tests/core/network/test_net_stream.py::test_net_stream_read_after_remote_closed - BaseExceptionGroup: Exceptions from Trio nursery (1 sub-exception)
FAILED tests/core/network/test_net_stream.py::test_net_stream_read_after_remote_closed_and_reset - ExceptionGroup: Exceptions from Trio nursery (1 sub-exception)

```

Going to take a look later

seetadev commented:
Awesome work @paschal533 and @acul71  — really appreciate you both jumping in and getting those failing tests fixed! 🎯 It's great to see Yamux making solid progress.

I saw there are just some tests still failing now. When you have some time, could you take another look and try to resolve these as well? Please ping me and @pacrob once completed.

If possible, it would be fantastic to get Yamux ready for review before Thursday. That way we can keep everything moving smoothly and start lining up next steps. Of course, if you run into anything tricky or blocking, feel free to flag it — happy to help out wherever needed.

Thanks again for all the hard work — we're really close now! 🚀

paschal533 commented:
Hi @acul71, Thank you so much for your help in fixing those test and reviews. 

@seetadev and @pacrob, we have fixed all the tests and they are all passing now. Need your review.

paschal533 commented:
> @paschal533 All the warnings during the documentation build process are related to the error: No module named 'libp2p.stream_muxer.yamux'. I believe this is occurring because the **init**.py file is missing in the yamux directory. As a result, it is not recognized as a Python module, which prevents other modules from importing it.

Hi Summajeet,

Nice catch on the missing `__init__.py` in the `yamux` directory. I agree that this is likely causing the `No module named 'libp2p.stream_muxer.yamux'` errors in the documentation build. It seems like a straightforward fix, and I plan to add the `__init__.py` file after getting reviews from Paul and Manu.

@pacrob, could you advise whether the `__init__.py` in `libp2p/stream_muxer/yamux/` should be empty, like in the `mplex` folder and others, or if we should import specific classes (e.g., `Yamux`) for consistency?

Thanks again, Summajeet, for pointing this out. I really appreciate your insight.

pacrob commented:
> @pacrob, could you advise whether the `__init__.py` in `libp2p/stream_muxer/yamux/` should be empty, like in the `mplex` folder and others, or if we should import specific classes for consistency?

I like adding the `__init__.py` and including an `all`  to define specifically what you want to be exported. It is clearer and cleaner.

acul71 commented on docs/libp2p.stream_muxer.yamux.rst:
```diff
@@ -0,0 +1,7 @@
+libp2p.stream\_muxer.yamux
+=========================
```
Add a `=` so that is matches the `libp2p.stream\_muxer.yamux` length
Like this
```
libp2p.stream\_muxer.yamux
==========================
```

paschal533 replied:
Thank you, Luca

paschal533 commented:
> > @pacrob, could you advise whether the `__init__.py` in `libp2p/stream_muxer/yamux/` should be empty, like in the `mplex` folder and others, or if we should import specific classes for consistency?
> 
> I like adding the `__init__.py` and including an `all` to define specifically what you want to be exported. It is clearer and cleaner.

Alright, noted. Thank you, Paul.

paschal533 commented:
Increasing trio.sleep from 0.01s to 0.1s to stabilize connections in CI’s slower environment would fix `test_notify` failure in `tests/core/network/test_notify.py`. What do you think @acul71, @seetadev, @sumanjeet0012, @pacrob ?

acul71 commented:
@paschal533 Great work Paschal you make all tests working!

Now we have to find a mechanism to select the multiplexer (`mplex`, `yamux`) and see that it works.

acul71 commented:
Hi @paschal533 
I've seen that `NOISE` is now the default `multiplexer`, but this is causing errors, better comment out or cancel it as default, and then work on this later, I will open a new issue soon.

# Noise Protocol Error and Fix

## The Error

After implementing Yamux as the default multiplexer, examples started failing with the following error:

```
NoiseValueError: Invalid length of private_bytes! Should be 32
```

This error occurs during the Noise protocol security handshake, specifically when trying to use the private key for encryption.

## Root Cause

The issue stems from how the Noise protocol was being initialized in the default security transport setup. The code was using the same key pair for both:
1. libp2p identity (typically RSA or Secp256k1)
2. Noise protocol encryption (which requires Ed25519)

The Noise protocol specifically requires an Ed25519 key pair for its encryption, but we were incorrectly using the libp2p identity key pair (which could be RSA or Secp256k1) for this purpose.

## Current Configuration

The current security transport setup in `libp2p/__init__.py` is:

```python
# Current configuration
secure_transports_by_protocol: Mapping[TProtocol, ISecureTransport] = sec_opt or {
    NOISE_PROTOCOL_ID: NoiseTransport(key_pair, noise_privkey=key_pair.private_key),
    TProtocol(PLAINTEXT_PROTOCOL_ID): InsecureTransport(key_pair),
    TProtocol(secio.ID): secio.Transport(key_pair),
}
```

This configuration:
1. Uses Noise as the primary security protocol
2. Uses the same key pair for both libp2p identity and Noise protocol encryption
3. Includes PLAINTEXT and SECIO as fallback options
4. Is causing the error because Noise protocol requires an Ed25519 key pair

## Historical Context

Before Yamux became the default multiplexer, the upstream `libp2p/__init__.py` had a different security transport configuration:

```python
# Upstream configuration (before Yamux)
security_transports_by_protocol = sec_opt or {
    TProtocol(PLAINTEXT_PROTOCOL_ID): InsecureTransport(key_pair),
    TProtocol(secio.ID): secio.Transport(key_pair),
}
```

The upstream version:
1. Used SECIO as the default security protocol
2. Did not include Noise protocol in the default configuration
3. Only included PLAINTEXT and SECIO transports

## Proposed Fix

The solution is to use separate key pairs for these two purposes:
1. Keep using the provided key pair (or generate a new RSA key pair) for libp2p identity
2. Generate a new Ed25519 key pair specifically for Noise protocol encryption

The proposed fix would be implemented in `libp2p/__init__.py` by modifying the default security transport setup:

```python
# Before
secure_transports_by_protocol = sec_opt or {
    NOISE_PROTOCOL_ID: NoiseTransport(key_pair, noise_privkey=key_pair.private_key),
    # ...
}

# After
secure_transports_by_protocol = sec_opt or {
    NOISE_PROTOCOL_ID: NoiseTransport(
        key_pair,
        noise_privkey=create_ed25519_key_pair().private_key,
    ),
    # ...
}
```

## Expected Benefits

1. **Backward Compatibility**: No changes required to existing examples or code that uses the default host creation
2. **Correct Key Types**: Each protocol would use the appropriate key type for its purpose
3. **Security**: Would maintain proper separation between identity and encryption keys
4. **Modern Protocol Support**: Would enable Noise protocol while maintaining compatibility with older protocols

## Testing Plan

The fix needs to be tested with:
- The identify example
- Other examples that were previously failing
- Both Yamux and Mplex multiplexers
- Various key pair types (RSA, Secp256k1) for libp2p identity

## Note

This proposed fix aims to ensure that the Noise protocol has the correct key type (Ed25519) for encryption while maintaining the existing behavior for libp2p identity. The examples should be able to continue using the default host creation without modification.

The transition from SECIO to Noise as the default security protocol, combined with the switch to Yamux as the default multiplexer, represents a significant upgrade in the security and performance capabilities of the library. However, this fix needs to be thoroughly tested before being implemented.

paschal533 commented:
> Hi @paschal533 @seetadev @pacrob I've seen that `NOISE` is now the default `protocol security`, but this is causing errors, better comment out or cancel it as default, and then work on this later, I will investigate more, and open a new issue soon. This is a draft version of the new Issue I will open
> 
> # Noise Protocol Error and Fix
> ## The Error
> After implementing Yamux as the default multiplexer, examples started failing with the following error:
> 
> ```
> NoiseValueError: Invalid length of private_bytes! Should be 32
> ```
> 
> This error occurs during the Noise protocol security handshake, specifically when trying to use the private key for encryption.
> 
> ## Root Cause
> The issue stems from how the Noise protocol was being initialized in the default security transport setup. The code was using the same key pair for both:
> 
> 1. libp2p identity (typically RSA or Secp256k1)
> 2. Noise protocol encryption (which requires Ed25519)
> 
> The Noise protocol specifically requires an Ed25519 key pair for its encryption, but we were incorrectly using the libp2p identity key pair (which could be RSA or Secp256k1) for this purpose.
> 
> ## Current Configuration
> The current security transport setup in `libp2p/__init__.py` is:
> 
> ```python
> # Current configuration
> secure_transports_by_protocol: Mapping[TProtocol, ISecureTransport] = sec_opt or {
>     NOISE_PROTOCOL_ID: NoiseTransport(key_pair, noise_privkey=key_pair.private_key),
>     TProtocol(PLAINTEXT_PROTOCOL_ID): InsecureTransport(key_pair),
>     TProtocol(secio.ID): secio.Transport(key_pair),
> }
> ```
> 
> This configuration:
> 
> 1. Uses Noise as the primary security protocol
> 2. Uses the same key pair for both libp2p identity and Noise protocol encryption
> 3. Includes PLAINTEXT and SECIO as fallback options
> 4. Is causing the error because Noise protocol requires an Ed25519 key pair
> 
> ## Historical Context
> Before Yamux became the default multiplexer, the upstream `libp2p/__init__.py` had a different security transport configuration:
> 
> ```python
> # Upstream configuration (before Yamux)
> security_transports_by_protocol = sec_opt or {
>     TProtocol(PLAINTEXT_PROTOCOL_ID): InsecureTransport(key_pair),
>     TProtocol(secio.ID): secio.Transport(key_pair),
> }
> ```
> 
> The upstream version:
> 
> 1. Used SECIO as the default security protocol
> 2. Did not include Noise protocol in the default configuration
> 3. Only included PLAINTEXT and SECIO transports
> 
> ## Proposed Fix
> The solution is to use separate key pairs for these two purposes:
> 
> 1. Keep using the provided key pair (or generate a new RSA key pair) for libp2p identity
> 2. Generate a new Ed25519 key pair specifically for Noise protocol encryption
> 
> The proposed fix would be implemented in `libp2p/__init__.py` by modifying the default security transport setup:
> 
> ```python
> # Before
> secure_transports_by_protocol = sec_opt or {
>     NOISE_PROTOCOL_ID: NoiseTransport(key_pair, noise_privkey=key_pair.private_key),
>     # ...
> }
> 
> # After
> secure_transports_by_protocol = sec_opt or {
>     NOISE_PROTOCOL_ID: NoiseTransport(
>         key_pair,
>         noise_privkey=create_ed25519_key_pair().private_key,
>     ),
>     # ...
> }
> ```
> 
> ## Expected Benefits
> 1. **Backward Compatibility**: No changes required to existing examples or code that uses the default host creation
> 2. **Correct Key Types**: Each protocol would use the appropriate key type for its purpose
> 3. **Security**: Would maintain proper separation between identity and encryption keys
> 4. **Modern Protocol Support**: Would enable Noise protocol while maintaining compatibility with older protocols
> 
> ## Testing Plan
> The fix needs to be tested with:
> 
> * The identify example
> * Other examples that were previously failing
> * Both Yamux and Mplex multiplexers
> * Various key pair types (RSA, Secp256k1) for libp2p identity
> 
> ## Note
> This proposed fix aims to ensure that the Noise protocol has the correct key type (Ed25519) for encryption while maintaining the existing behavior for libp2p identity. The examples should be able to continue using the default host creation without modification.
> 
> The transition from SECIO to Noise as the default security protocol, combined with the switch to Yamux as the default multiplexer, represents a significant upgrade in the security and performance capabilities of the library. However, this fix needs to be thoroughly tested before being implemented.
> 
> ##########################
> 
> The test coverage of `NOICE` appears to be comprehensive, covering: Basic functionality Connection establishment Data exchange Protocol integration Interoperability However, I notice that the tests don't explicitly verify the key type requirements for Noise protocol. This might explain why the issue with using non-Ed25519 keys wasn't caught earlier.

Thank you, Luca. I will work on this today and give my update by evening.

paschal533 commented:
> Hi @paschal533 @seetadev @pacrob I've seen that `NOISE` is now the default `protocol security`, but this is causing errors, better comment out or cancel it as default, and then work on this later, I will investigate more, and open a new issue soon. This is a draft version of the new Issue I will open
> 
> # Noise Protocol Error and Fix
> ## The Error
> After implementing Yamux as the default multiplexer, examples started failing with the following error:
> 
> ```
> NoiseValueError: Invalid length of private_bytes! Should be 32
> ```
> 
> This error occurs during the Noise protocol security handshake, specifically when trying to use the private key for encryption.
> 
> ## Root Cause
> The issue stems from how the Noise protocol was being initialized in the default security transport setup. The code was using the same key pair for both:
> 
> 1. libp2p identity (typically RSA or Secp256k1)
> 2. Noise protocol encryption (which requires Ed25519)
> 
> The Noise protocol specifically requires an Ed25519 key pair for its encryption, but we were incorrectly using the libp2p identity key pair (which could be RSA or Secp256k1) for this purpose.
> 
> ## Current Configuration
> The current security transport setup in `libp2p/__init__.py` is:
> 
> ```python
> # Current configuration
> secure_transports_by_protocol: Mapping[TProtocol, ISecureTransport] = sec_opt or {
>     NOISE_PROTOCOL_ID: NoiseTransport(key_pair, noise_privkey=key_pair.private_key),
>     TProtocol(PLAINTEXT_PROTOCOL_ID): InsecureTransport(key_pair),
>     TProtocol(secio.ID): secio.Transport(key_pair),
> }
> ```
> 
> This configuration:
> 
> 1. Uses Noise as the primary security protocol
> 2. Uses the same key pair for both libp2p identity and Noise protocol encryption
> 3. Includes PLAINTEXT and SECIO as fallback options
> 4. Is causing the error because Noise protocol requires an Ed25519 key pair
> 
> ## Historical Context
> Before Yamux became the default multiplexer, the upstream `libp2p/__init__.py` had a different security transport configuration:
> 
> ```python
> # Upstream configuration (before Yamux)
> security_transports_by_protocol = sec_opt or {
>     TProtocol(PLAINTEXT_PROTOCOL_ID): InsecureTransport(key_pair),
>     TProtocol(secio.ID): secio.Transport(key_pair),
> }
> ```
> 
> The upstream version:
> 
> 1. Used SECIO as the default security protocol
> 2. Did not include Noise protocol in the default configuration
> 3. Only included PLAINTEXT and SECIO transports
> 
> ## Proposed Fix
> The solution is to use separate key pairs for these two purposes:
> 
> 1. Keep using the provided key pair (or generate a new RSA key pair) for libp2p identity
> 2. Generate a new Ed25519 key pair specifically for Noise protocol encryption
> 
> The proposed fix would be implemented in `libp2p/__init__.py` by modifying the default security transport setup:
> 
> ```python
> # Before
> secure_transports_by_protocol = sec_opt or {
>     NOISE_PROTOCOL_ID: NoiseTransport(key_pair, noise_privkey=key_pair.private_key),
>     # ...
> }
> 
> # After
> secure_transports_by_protocol = sec_opt or {
>     NOISE_PROTOCOL_ID: NoiseTransport(
>         key_pair,
>         noise_privkey=create_ed25519_key_pair().private_key,
>     ),
>     # ...
> }
> ```
> 
> ## Expected Benefits
> 1. **Backward Compatibility**: No changes required to existing examples or code that uses the default host creation
> 2. **Correct Key Types**: Each protocol would use the appropriate key type for its purpose
> 3. **Security**: Would maintain proper separation between identity and encryption keys
> 4. **Modern Protocol Support**: Would enable Noise protocol while maintaining compatibility with older protocols
> 
> ## Testing Plan
> The fix needs to be tested with:
> 
> * The identify example
> * Other examples that were previously failing
> * Both Yamux and Mplex multiplexers
> * Various key pair types (RSA, Secp256k1) for libp2p identity
> 
> ## Note
> This proposed fix aims to ensure that the Noise protocol has the correct key type (Ed25519) for encryption while maintaining the existing behavior for libp2p identity. The examples should be able to continue using the default host creation without modification.
> 
> The transition from SECIO to Noise as the default security protocol, combined with the switch to Yamux as the default multiplexer, represents a significant upgrade in the security and performance capabilities of the library. However, this fix needs to be thoroughly tested before being implemented.
> 
> ##########################
> 
> The test coverage of `NOICE` appears to be comprehensive, covering: Basic functionality Connection establishment Data exchange Protocol integration Interoperability However, I notice that the tests don't explicitly verify the key type requirements for Noise protocol. This might explain why the issue with using non-Ed25519 keys wasn't caught earlier.

Thanks Luca for catching this Noise protocol issue. You're right on target with the key type problem. I totally missed that Noise needs Ed25519 keys specifically.

I'll implement your fix by separating the keys:
- Keep the identity key pair as is
- Generate a proper Ed25519 key for Noise

Here's what I'm thinking:

```python
from libp2p.crypto.ed25519 import create_new_key_pair as create_ed25519_key_pair

secure_transports_by_protocol = sec_opt or {
    NOISE_PROTOCOL_ID: NoiseTransport(
        key_pair,
        noise_privkey=create_ed25519_key_pair().private_key,
    ),
    TProtocol(secio.ID): secio.Transport(key_pair),
    TProtocol(PLAINTEXT_PROTOCOL_ID): InsecureTransport(key_pair),
}
```

I will also check with different key types and both multiplexers to make sure everything's solid.

Really appreciate the detailed breakdown you provided... saved me a ton of debugging time. Let me know if you think this approach works or if you have any other thoughts.

acul71 commented:
Hi @paschal533 
I've done some test and example fixing.
Please review this [PR fix: fixing failing tests and examples with yamux and noise](https://github.com/paschal533/py-libp2p/pull/2)

The tests and lint are ok, but In the code there is still some debugging enabled, and comment I have to strip, I will do it before the py-libp2p meeting

acul71 commented:
Hi @paschal533 also we have to check for configuring how to select `MPLEX` as multistreamer, since now `Yamux` is the default one, but we have to be sure `mplex` work for backward compatibility with `ethereum` old clients. 
Can you suggest something?

paschal533 commented:
> Hi @paschal533 also we have to check for configuring how to select `MPLEX` as multistreamer, since now `Yamux` is the default one, but we have to be sure `mplex` work for backward compatibility with `ethereum` old clients. Can you suggest something?

I think there are a few simple ways we could handle this:

First, maybe add an environment variable to let people choose their preferred muxer:

```python
# In libp2p/__init__.py
default_muxer = os.environ.get("LIBP2P_DEFAULT_MUXER", "YAMUX").upper()

# Then use this to decide the order
muxer_transports_by_protocol = muxer_opt or (
    {MPLEX_PROTOCOL_ID: Mplex, cast(TProtocol, "/yamux/1.0.0"): Yamux}
    if default_muxer == "MPLEX" else
    {cast(TProtocol, "/yamux/1.0.0"): Yamux, MPLEX_PROTOCOL_ID: Mplex}
)
```

We could also add a simple helper function:

```python
def set_default_muxer(muxer_name: str) -> None:
    """Pick your multiplexer: 'YAMUX' or 'MPLEX'"""
    global default_muxer
    if muxer_name.upper() not in ["YAMUX", "MPLEX"]:
        raise ValueError(f"Unknown muxer: {muxer_name}. Use 'YAMUX' or 'MPLEX'.")
    default_muxer = muxer_name.upper()
```

Maybe we could even auto-detect Ethereum clients? Something like:

```python
async def detect_ethereum_client(peer_id: ID) -> bool:
    # Some logic to detect Ethereum clients
    pass

# And then in the connection flow:
if await detect_ethereum_client(peer_id):
    # Use MPLEX for this connection
```

We will also add some quick docs showing Ethereum devs how to use this:

```python
# For Ethereum stuff:
import os
os.environ["LIBP2P_DEFAULT_MUXER"] = "MPLEX"
import libp2p
```

It'd also be good to test with actual Ethereum clients to make sure everything works.

This way Yamux can still be the default for modern stuff, but MPLEX is there when people need it

seetadev commented:
@paschal533 and @acul71 : Great work and wonderful contribution. Please continue research and development discussion on yamux + mplex at this web page: please visit https://github.com/libp2p/py-libp2p/discussions/582. I am also adding some resources that we should check from an interop perspective. We will discuss with @pacrob about it today.

Kindly keep the PR page for code commits and refinement. Appreciate your great initiative and progress.

acul71 commented:
@seetadev 
@paschal533 Hi Paschal,
I hope you’re doing well. I wanted to follow up because I see that one of my pull requests (#2: fixing failing tests and examples) has been open for about two weeks now, and PRs #3 and #4 have also been pending for over a week. Meanwhile, PR #5 is brand new.

Would you mind taking a look at those older PRs when you have a moment? If there’s anything I can clarify or update to help get them merged, just let me know.

Thanks a lot for your help!
Best,
Luca

paschal533 commented:
> @paschal533 and @acul71 : Great work. The branch on which you are working has conflicts that needs resolution.
> 
> Please visit tests/conftest.py.
> 
> Looking forward to a resolution and having a fully functional yamux + mplex. Please tag @pacrob , me and @dhuseby once the merge conflicts have been resolved.
> 
> On the same note, wish to share that we have already made good progress on interop with rust-libp2p. Thanks to 2 new PLDG contributors. Looking forward to yamux integration.

Okay Manu. I will resolve the conflict this morning. Thanks

seetadev commented:
@paschal533 and @acul71  : Please do share developer documentation and a demo video for using both yamux and mplex together. Focus more on yamux functionality. Also, share how developers can use the stack for interop tests.

acul71 commented:
> Hi @pacrob and @seetadev, everything is ready now 🎉
> 
> Thanks @acul71 For your incredible contributions.

Hi @paschal533  just one small thing, can you change the PR title, I've seen in discord that is creating confusion. Something like "feat: Adding Yamux as default multiplexer, keeping Mplex as fallback"

pacrob reviewed (approved):
looks good!
